openvpn/build/build.py

```diff
@@ -152,131 +281,476 @@ def on_error(func, path, exc_info):
+            os.path.join('out32dll', 'ssleay32.dll'),
+            os.path.join(STAGE_DIR, 'x64'))
+
+
+def build_oqs_openssl_unix(test_build=False):
+    '''
+    Build openssl for linux/unix
+    '''
+    git_restore(OPENSSL_OQS_REPO_DIRNAME)
+
+    with chdir(OPENSSL_OQS_REPO_DIRNAME):
+        target_system = run_command_capture(
+            ['/bin/sh', OPENSSL_CONFIG_PLATFORM_SCRIPT])
+        target_system = target_system.strip()
+        run_command([
+            'sed', '-i org', '1s|^|#!/usr/bin/env perl\\\n#|', 'Configure'])
+        run_command([
+            './Configure', 'shared',
+            '--prefix={PREFIX}'.format(PREFIX=PREFIX),
+            target_system])
+        run_command(['make'])
+        if test_build:
+            run_command(['make', 'test'])
+        run_command([
+            'make', 'install',
+            'INSTALL_PREFIX={STAGE_DIR}'.format(STAGE_DIR=STAGE_DIR),
+            'INSTALL_TOP="/"'])
+
+
+#
+# OpenVPN
+#
+
+
+def build_openvpn():
+    '''
+    Build openvpn on linux/unix/windows
+    '''
+
+    git_restore(OPENVPN_REPO_DIRNAME)
+
+    with chdir(OPENVPN_REPO_DIRNAME):
+        run_command(['autoreconf', '-i', '-f', '-v'])
+        lib_dir = os.path.abspath(
+            os.path.join(STAGE_DIR, PREFIX[1:], 'lib'))
+        include_dir = os.path.abspath(
+            os.path.join(STAGE_DIR, PREFIX[1:], 'include'))
+
+        if not os.path.exists(lib_dir):
+            die(
+                'Did not find lib_dir: {lib_dir} directory.'.format(
+                    lib_dir=lib_dir))
+        if not os.path.exists(include_dir):
+            die(
+                'Did not find include_dir: {include_dir} directory.'.format(
+                    include_dir=include_dir))
+
+        extra_openssl_libs = ''
+        if BUILD_TARGET_OS == 'Linux':
+            extra_openssl_libs = '-Wl,-rpath={prefix}/lib'.format(
+                prefix=PREFIX)
+
+        run_env = dict(
+            OPENSSL_LIBS=(
+                '-L{lib_dir} {extra_openssl_libs} -lssl -lcrypto'.format(
+                    lib_dir=lib_dir,
+                    extra_openssl_libs=extra_openssl_libs)),
+            OPENSSL_CFLAGS='-I{include_dir}'.format(include_dir=include_dir),
+            PKCS11_HELPER_CFLAGS='',
+            PKCS11_HELPER_LIBS='')
+
+        if BUILD_TARGET_OS == 'Windows':
+            run_env['CHOST'] = 'i686-w64-mingw32'
+            run_env['CBUILD'] = 'i686-pc-linux-gnu'
+
+        run_command([
+            './configure',
+            '--prefix={prefix}'.format(prefix=PREFIX),
+            '--disable-plugins',
+            '--with-special-build=pq-openvpn'],
+            environment=run_env)
+        run_command(['make'], environment=run_env)
+        run_command([
+            'make', 'install',
+            'DESTDIR={destdir}'.format(destdir=STAGE_DIR)])
+
+
+def pack_images_windows():
+    '''
+    Pack the redistributable image for windows
+    '''
+    with chdir(OPENVPN_BUILD_REPO_DIRNAME):
```

Comment:
I'm getting an odd error at this line when I run: python build.py --target=Windows --skip-test

Traceback (most recent call last):
  File "build.py", line 754, in <module>
    pack_images()
  File "build.py", line 627, in pack_images
    pack_images_windows()
  File "build.py", line 522, in pack_images_windows
    with chdir(OPENVPN_BUILD_REPO_DIRNAME):
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "build.py", line 120, in chdir
    os.chdir(path)

Before this, it does appear to configure and make in the build/build/openvpn-2.4.4 subtree. Can you investigate?

Reply:
OPENVPN_BUILD_REPO_DIRNAME is a relative name that only exists in SOURCES_DIR after download_remote_repositories. I think the windows build needs love.
but when you got to this point, the build_openvpn was success.
Rewrite pack_images_windows is required when result tree is reviewed